vnext/Desktop/WinHTTPWebSocketResource.cpp

```diff
@@ -0,0 +1,447 @@
+    m_requestHandle,                          // hInternet
+    Callback,                                 // lpfnInternetCallback
+    WINHTTP_CALLBACK_FLAG_ALL_NOTIFICATIONS,  // dwNotificationFlags
+    0                                         // dwReserved
+  ) == WINHTTP_INVALID_STATUS_CALLBACK)
+  {
+    m_errorStatus = GetLastError();
+    throw std::exception("Can not create instance. Invalid callback.");
+  }
+
+  BOOL fStatus = FALSE; //TODO: Make member?
+  fStatus = WinHttpSetOption(
+    m_requestHandle,                      // hInternet
+    WINHTTP_OPTION_UPGRADE_TO_WEB_SOCKET, // dwOption
+    NULL,                                 // lpBuffer
+    0                                     // dwBufferLength
+  );
+
+  if (!fStatus)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error upgrading protocol", ErrorType::Connection });
+    }
+  }
+
+  // Perform handshake
+  fStatus = WinHttpSendRequest(
+    m_requestHandle,                // hRequest
+    WINHTTP_NO_ADDITIONAL_HEADERS,  // lpszHeaders
+    0,                              // dwHeadersLength
+    NULL,                           // lpOptional
+    0,                              // dwOptionalLength
+    0,                              // dwTotalLength
+    (DWORD_PTR)this                 // dwContext
+  );
+  if (!fStatus)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error sending handshake", ErrorType::Handshake });
+    }
+  }
+  fStatus = WinHttpReceiveResponse(m_requestHandle, /*lpReserved*/ 0);
+  if (!fStatus)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error receiving handshake", ErrorType::Handshake });
+    }
+  }
+
+  // Complete upgrade. Accept only HTTP status code 101.
+  m_webSocketHandle = WinHttpWebSocketCompleteUpgrade(m_requestHandle, /*pContext*/ NULL);
+  if (m_webSocketHandle == NULL)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error completing protocol upgrade", ErrorType::Connection });
+    }
+  }
+
+  // Connection succeeded.
+  if (m_connectHandler)
+  {
+    m_connectHandler();
+  }
+}
+
+void WinHTTPWebSocketResource::Ping() {}
+
+void WinHTTPWebSocketResource::Send(const string& message)
+{
+}
+
+void WinHTTPWebSocketResource::SendBinary(const string& base64String)
+{
+}
+
+void WinHTTPWebSocketResource::Close(CloseCode code, const string& reason)
+{
+  auto result = WinHttpWebSocketShutdown(
+    m_webSocketHandle,                                         // hWebSocket
+    static_cast<USHORT>(code),                                 // usStatus
+    static_cast<LPWSTR>(_wcsdup(Utf8ToUtf16(reason).c_str())), // pvReason
+    static_cast<DWORD>(reason.length())                        // dwReasonLength
+  );
+  switch (result)
+  {
+  case ERROR_IO_PENDING:
+    GetLastError();
+    if (m_errorHandler)
```

Comment:
what's this doing?

Reply:
Nothing really. It indicates the close operation will finish asynchronously. It doesn't need to trigger an error.

Fixing.